### PR TITLE
[ndarray] Infer types from input `data` array

### PR DESCRIPTION
**Pull request by axelboc**
⚠️ **Breaking change!** ⚠️ 

The generic type of the `ndarray` function now refers to the type of the input array instead of the type of the values contained in the input array: `ndarray<D>(data: D)`. This is a breaking change for cases where the generic is specified explicitly. If the generic is not specified, it is inferred from the input array.

This change brings some major benefits, notably:

1. The input array exposed through `arr.data` no longer "loses" its type.
2. `arr.dtype` can now be inferred to the correct string.
3. In most cases, the generic no longer has to be specified when calling `ndarray()`, even when dealing with input values of another type than `number` -- everything gets inferred properly from the type of the input array.
4. It's no longer possible to specify a generic that is incompatible with the input array.

```tsx
// BEFORE
const arr = ndarray([0, 1]);
const data: number[] = arr.data as number[]; // `arr.data` has type `number[] | Int8Array | ...
const dtype = arr.dtype; // 'int8' | 'int16' | ...

// AFTER
const arr = ndarray([0, 1]); // no breaking change here, everything just works
const data: number[] = arr.data; // no cast required
const dtype = arr.dtype; // 'array'
```

```tsx
// BEFORE
const arr = ndarray(Float32Array.from([0, 1]));
const data: ArrayBuffer = (arr.data as Float32Array).buffer;
const dtype = arr.dtype; // 'int8' | 'int16' | ...

// AFTER
const arr = ndarray(Float32Array.from([0, 1])); // no breaking change here, everything just works
const data: ArrayBuffer = arr.data.buffer; // no cast required
const dtype = arr.dtype; // 'float32'
```

```tsx
// BEFORE
const arr = ndarray<string>(['foo', 'bar']);
const data: string[] = arr.data as string[];

// AFTER
const arr = ndarray(['foo', 'bar']); // BREAKING CHANGE: generic must be removed or replaced with `string[]`
const data: string[] = arr.data; // no cast required
```

```tsx
// BEFORE
function doSomething(arr: NdArray<string>) { ... }

// AFTER
function doSomething(arr: NdArray<string[]>) { ... } // BREAKING CHANGE: generic must be replaced with `string[]`
```

```tsx
// BEFORE
const arr = ndarray<string>(Float32Array.from([0,1])); // no error, even though a `Float32Array` cannot possibly contain strings

// AFTER
const arr = ndarray<string[]>(Float32Array.from([0,1])); // error
//                            ^ Argument of type 'Float32Array' is not assignable to parameter of type 'string[]'
```

Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [x] Test the change in your own code. (Compile and run.)
- [x] [Add or edit tests](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#my-package-teststs) to reflect the change.
- [x] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [x] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [x] [Run `npm test <package to test>`](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#running-tests).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: https://github.com/scijs/ndarray/blob/master/ndarray.js
- [x] ~If this PR brings the type definitions up to date with a new version of the JS library, update the version number in the header.~

### CODE REVIEW

File: types/ndarray/index.d.ts [hunk 1]
```diff
@@ -6,60 +6,67 @@
...
     shape?: number[],
     stride?: number[],
     offset?: number,
-): ndarray.NdArray<T>;
+): ndarray.NdArray<D>;
 
 declare namespace ndarray {
-    interface NdArray<T = number> {
-        data: Data<T>;
+    interface NdArray<D extends Data = Data<number>> {
```
Comment (axelboc) [inline types/ndarray/index.d.ts:17]: The default type `Data<number>` ensures backwards compatibility when using `NdArray` as an argument or return type without a generic.

File: types/ndarray/index.d.ts [hunk 2]
```diff
@@ -6,60 +6,67 @@
...
-        | "uint8"
-        | "uint16"
-        | "uint32"
-        | "float32"
-        | "float64"
-        | "array"
-        | "uint8_clamped"
-        | "buffer"
-        | "generic";
+    type Value<D extends Data> = D extends Array<infer T> ? T : number;
```
Comment (axelboc) [inline types/ndarray/index.d.ts:49]: `Value<D>` infers `number` when given a typed array.

File: types/ndarray/index.d.ts [hunk 3]
```diff
@@ -6,60 +6,67 @@
...
         dimension: number;
-        get(...args: number[]): T;
-        set(...args: number[]): T;
-        index(...args: number[]): T;
-        lo(...args: number[]): NdArray<T>;
-        hi(...args: number[]): NdArray<T>;
-        step(...args: number[]): NdArray<T>;
-        transpose(...args: number[]): NdArray<T>;
-        pick(...args: Array<number | null>): NdArray<T>;
-        reshape(...shapes: number[]): NdArray<T>;
```
Comment (axelboc) [inline types/ndarray/index.d.ts:34]: I removed `reshape` as I don't believe it exists any more: https://github.com/scijs/ndarray/blob/master/ndarray.js

File: types/ndarray/index.d.ts [hunk 4]
```diff
@@ -6,60 +6,67 @@
...
+        ? 'uint8_clamped'
+        : D extends Uint16Array
+        ? 'uint16'
+        : D extends Uint32Array
+        ? 'uint32'
+        : D extends Float32Array
+        ? 'float32'
+        : D extends Float64Array
+        ? 'float64'
+        : 'array';
```
Comment (axelboc) [inline types/ndarray/index.d.ts:69]: I've removed `'buffer'` and `'generic'` as this package forbids creating an `ndarray` with anything else than an array or typed array.

File: types/numjs/index.d.ts [hunk 5]
```diff
@@ -9,7 +9,7 @@ import { Data, DataType, NdArray as BaseNdArray } from "ndarray";
 
 export type NdType<T> = DataType | Data<T>;
 
-export interface NdArray<T = number> extends BaseNdArray<T> {
+export interface NdArray<T = number> extends BaseNdArray<Data<T>> {
```
Comment (axelboc) [inline types/numjs/index.d.ts:12]: This is the only breaking change I could find in other types packages. Users of `numjs` should neither be impacted by any of the breaking changes ... nor benefit from any of the improvements until the generic `T = number` is replace. @taoqf @mattmm3d

**Review by axelboc — COMMENTED** [review]
Any feedback welcome 😉

File: types/ndarray/index.d.ts (hunk 3 above)
Comment (amcasey) [reply types/ndarray/index.d.ts:34]: The version number of this package is consistent with the version number of `ndarray` in which that function was removed?

**Comment by amcasey** [comment]
> ```tsx
> // BEFORE
> const arr = ndarray([0, 1]);
> const data: number[] = arr.data as number[]; // `arr.data` has type `number[] | Int8Array | ...
> const dtype = arr.dtype; // 'int8' | 'int16' | ...
> 
> // AFTER
> const arr = ndarray([0, 1]); // no breaking change here, everything just works
> const data: number[] = arr.data; // no cast required
> const dtype = arr.dtype; // 'array'
> ```

So the old `dtype` was just totally wrong?  I'm having trouble deducing from these example what `dtype` is supposed to do.  From the docs, it sounds like it's supposed to give the type used for storing the array, so returning an element type seems like it was always wrong?

**Comment by amcasey** [comment]
Seems like a good change to me, but it would be nice to hear from another stakeholder.

File: types/ndarray/index.d.ts (hunk 3 above)
Comment (axelboc) [reply types/ndarray/index.d.ts:34]: I don't believe this method was ever part of `ndarray`. I traced back its addition in `@types/ndarray` to this commit by @taoqf:  https://github.com/DefinitelyTyped/DefinitelyTyped/pull/18343/files. `numjs` augments `ndarray` with this method: https://github.com/nicolaspanel/numjs/blob/master/src/ndarray.js#L241, so the method should be typed in `@types/numjs` instead.

Comment (axelboc) [reply types/ndarray/index.d.ts:34]: Done! Thanks for getting me to double check this. 💯

**Comment by orta** [comment]
Just also popping by to say that we do want someone else to give an opinion on this change

**Comment by elibarzilay** [comment]
Making it possible to merge since it's been a while and there's no objections...  (Would be good still to get some people who use these.)

**Comment by axelboc** [comment]
Ready to merge

**Pull request merged**